### DISCUSSION TITLE
issue: useWatch / watch has wrong values when using form array with dnd kit

### DISCUSSION THREAD

**Discussion by alex-collibra** [react-hook-form/react-hook-form#13513] [Q&A]
### Version Number

7.43.9

### Codesandbox/Expo snack

[Codesandbox reproduction](https://codesandbox.io/p/devbox/nz79q4?file=%2Fsrc%2FApp.tsx%3A188%2C22&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwet8kzs00063b6l7ks6d7li%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwet8kzr00023b6lvi520cpy%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwet8kzr00043b6ls15ikedd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwet8kzr00053b6lw2ggxl5n%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B49.42797935383304%252C50.57202064616696%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwet8kzr00023b6lvi520cpy%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwet8kzr00013b6l3gkznjqb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clweta0vr00023b6l2a8zpe7j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A4%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A4%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fstyles.css%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clwetdl0f00023b6ltvsw2a7a%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A9%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A9%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Ftailwind.config.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clweti2a900023b6l5l29ydbe%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A188%252C%2522startColumn%2522%253A22%252C%2522endLineNumber%2522%253A188%252C%2522endColumn%2522%253A22%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clwet8kzr00023b6lvi520cpy%2522%252C%2522activeTabId%2522%253A%2522clweti2a900023b6l5l29ydbe%2522%257D%252C%2522clwet8kzr00053b6lw2ggxl5n%2522%253A%257B%2522id%2522%253A%2522clwet8kzr00053b6lw2ggxl5n%2522%252C%2522activeTabId%2522%253A%2522clwet93pq00763b6lobx7h92e%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A5173%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clwet93pq00763b6lobx7h92e%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clwet8kzr00043b6ls15ikedd%2522%253A%257B%2522id%2522%253A%2522clwet8kzr00043b6ls15ikedd%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwet8kzr00033b6li6m3r1ri%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clwet8lo0000adbf8d97o2hnp%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clwet8w63005d3b6lg8svtgwj%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522pnpm%2520install%2520postcss%25408.4.38%2522%252C%2522id%2522%253A%2522clweten6g008e3b6lx7nmm28u%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522pnpm%2520install%2520autoprefixer%254010.4.19%2522%252C%2522id%2522%253A%2522clwetesik00a83b6lxz90f4pr%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clwet8w63005d3b6lg8svtgwj%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce

1. Open provided codesandbox
2. See there are 2 field there: one has conditional field for grade
3. Drag the line with two fields to swap it with another one.
There are two issues happening:
1. `useWatch` in `GradeField` returns `undefined` value for `additionalInformation` and renders the component, even though `type` === 'adult' and so that conditional field should be hidden completely.
2. `useWatch` in top lvl component shows unexpected values (`grade` property is now existing on both form rows, while it should only be existing for one row. The values in onSubmit are correct)

It seems like either `useWatch` runs earlier than expected or maybe there is a problem with `keys` (due to how drag overlay works)

Given that [dnd kit react](https://dndkit.com) is widely used  and they recommend using DragOverlay, it could be really helpful to have some examples / recommendations.

Thanks in advance 🙏 

### Expected behaviour

`useWatch` should show proper values.
`useWatch` should not return `undefined` for field which shouldn't even be rendered.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Pretty sure it's related to this: 

> useWatch's execution order matters, which means if you update a form value before the subscription is in place, then the value updated will be ignored.

https://react-hook-form.com/docs/usewatch

**Comment by alex-collibra** [comment]
I see. Using `useFormValues` hook as suggested in the docs and getting `type` from `useFormValues` return value instead of accessing it using `field` kind of works. 
The problem is that old values still shows up for a moment, where they shouldn't.

Before:
```
const useFormValues = () => {
  const { getValues } = useFormContext()

  return {
    ...useWatch(), // subscribe to form value updates

    ...getValues(), // always merge with latest form values
  }
}

// FormRow component

const FormRow = ({ index, field, isFromDragOverlay }) => {

  return (
      ...
      {field.type === 'student' && (
        <GradeField index={index} isFromDragOverlay={isFromDragOverlay} />
      )}
      ...
  )
}
```

After:
```
const FormRow = ({ index }) => {
  const values = useFormValues()
  const type = values.people[index].type

  return (
    ...
    {type === 'student' && <GradeField index={index} />} // <--- no need to pass isFromDragOverlay now or disable watch if field is from drag overlay - it will just work now
    ...
    )
}
```

**Comment by bluebill1049** [comment]
This PR may address this sort of issue https://github.com/react-hook-form/react-hook-form/pull/11522 as subscription can be established before react component.

**Comment by BrendanC23** [comment]
@bluebill1049 Can you explain how `useFormValues()` works? Is this accurate?

* `useWatch()` will re-trigger at the hook level (instead of the `useForm()` root level like `watch()`), but it doesn't always have the most up-to-date value because the update might happen before the subscription
* `getValues()` will have the up-to-date values

Why does `useFormValues()` return `{ ...useWatch(), ...getValues() }`? Why not just call `useWatch()` to ensure that the function runs when the subscription changes, and then return `getValues()` to get the latest values?

Is there any issues with doing something like this instead?

```
export function useWatchFormValues<
    TFieldValues extends FieldValues = FieldValues,
    TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
>(form: UseFormReturn<TFieldValues>, name: TFieldName) {
    useWatch({ control: form.control, exact: true, name: name }); // subscribe to form value updates
    const values = form.getValues(name);

    return values;
}

```

**Comment by ap0nia** [comment]
I think adding a key to the `DragOverlay` actually works! It seemed like everything outside the overlay had the correct values but an erroneus `FormRow` continued to be rendered even after the activeId was reset.

```tsx
<DragOverlay key={activeId}>
  {activeId ? (
    <FormRow {...getActiveChild()} isFromDragOverlay />
  ) : null}
</DragOverlay>
```

Based on the code [here](https://github.com/clauderic/dnd-kit/blob/00f749bc0cc3e6582f4f887f64c1f1de65ee0081/packages/core/src/components/DragOverlay/DragOverlay.tsx#L29), the overlay is indeed memo-ized and potentially the root cause.

> I see. Using `useFormValues` hook as suggested in the docs and getting `type` from `useFormValues` return value instead of accessing it using `field` kind of works. The problem is that old values still shows up for a moment, where they shouldn't.

In this instance, the previous active child continues to be rendered despite being moved. This lines up with your previous observation that the old values continue to persist there for some reason. It seems like this is likely due to dndkit and not react-hook-form.

See the updated sandbox [here](https://codesandbox.io/p/devbox/nz79q4).
